### PR TITLE
ARROW-4028: [Rust] Merge parquet-rs codebase

### PR DESCRIPTION
**Pull request by sunchao**
This imports parquet-rs source code into Apache Arrow Rust implementation. I include most of the source code except a few things such as `fuzz` and benchmarks. Thinking about adding them later.

The module hierarchy now looks like:
- arrow: all the arrow code
- parquet: all the parquet code (in future, parquet-arrow integration will live here)
- util: common util libraries shared between arrow and parquet (I'll try to move the utils from parquet to here in future).

### CODE REVIEW

File: rust/src/parquet/errors.rs
Comment (sadikovi) [inline rust/src/parquet/errors.rs]: We might need to fix formatting for this macro to use 4 spaces.

**Review by sadikovi — APPROVED** [review]
Thanks for the PR! Looks good. Do we still need parquet record reader? I am also wondering what will happen to our documentation release process.

**Comment by sunchao** [comment]
@sadikovi yes we should keep the parquet record reader since it is still useful. For documentation, we can release it on each new crate publish. We can combine with arrow's documentation and host it in a new place. Choices could be `docs.rs` or arrow.apache.org, or both. This needs some work on the travis CI.

File: rust/src/parquet/errors.rs
Comment (sunchao) [reply rust/src/parquet/errors.rs]: Good catch. Interestingly rustfmt didn't complain about this..

File: rust/benches/array_from_vec.rs
Comment (kszucs) [inline rust/benches/array_from_vec.rs]: @sunchao I wonder how could We avoid having this redundant `arrow::` namespace? 
What other structuring approaches do We have (e.g. a cargo workspace)? What are the advantages/disadvantages of them?

File: rust/examples/dynamic_types.rs
Comment (andygrove) [inline rust/examples/dynamic_types.rs]: The repeated namespace isn't ideal ... this is because we are building both arrow and parquet as a single crate? Maybe we should keep it as separate crates?

**Comment by wesm** [comment]
I think you'll want to remove the data files and use the ones in https://github.com/apache/parquet-testing if possible (you could need to contribute some more data files to that repo)

File: rust/examples/dynamic_types.rs
Comment (paddyhoran) [reply rust/examples/dynamic_types.rs]: What do we think about adding `parquet` alongside the `csv` as a sub module if building as a single crate?

Comment (sunchao) [reply rust/examples/dynamic_types.rs]: Making them separate crate adds extra dependency - say if I want to use a new feature in arrow for my parquet arrow reader, I have to wait for the next crate release, which is undesirable. 

Yes I'm fine by keeping `parquet` alongside `csv` and keep the `arrow` code top level as they are right now. This can solve the repeated namespace issue.

**Comment by sunchao** [comment]
> I think you'll want to remove the data files and use the ones in https://github.com/apache/parquet-testing if possible (you could need to contribute some more data files to that repo)

Got it. Let me try it. I think the CI already exposed the `PARQUET_TEST_DATA` env var so maybe we can just directly use it in unit tests. 

> The project (arrow) does not build out of the box for windows users as clang is now a dependency. The CI image includes it by default. Maybe add a note to the README regarding this with the following link. Obviously not a blocker for the PR... I can follow up with with proper instructions for windows users.

Got it. I'll update the README.

**Comment by wesm** [comment]
@sunchao I see that there are 7 contributors in https://github.com/sunchao/parquet-rs/graphs/contributors. Whose work is reflected in this patch? I think we only have CLAs for you, Ivan, and Andy

**Comment by sunchao** [comment]
@wesm the other authors are @gnieto, @xrl and @liurenjie1024. One other author's code is not included in this PR.

**Comment by wesm** [comment]
OK, the easiest thing from an IP clearance perspective is to make sure we have ICLAs on file for everyone

**Comment by wesm** [comment]
Needs rebase. I'm going to start the IP clearance vote today

**Comment by sunchao** [comment]
Sure. Rebased.

**Comment by wesm** [comment]
IP clearance vote: https://lists.apache.org/thread.html/7e4eca3584d948fb89beed6dbfd24cff01c7053eb275761960ae9704@%3Cgeneral.incubator.apache.org%3E. 

Can you open a JIRA and have this ready to merge in ~72 hours? Thanks

**Comment by sunchao** [comment]
Sure. I'll create the JIRA and rebase this PR later today.

**Review by wesm — APPROVED** [review]
+1. The IP clearance is done. Once the RAT issue is fixed this can be merged. Taking a look

**Comment by wesm** [comment]
@sunchao can you fix the broken build? It's complaining about the data files

**Comment by sunchao** [comment]
Sure. We need to update the `parquet-testing` git submodule.

**Comment by sunchao** [comment]
Thanks everyone!!

**Pull request closed without merging**
